atl-checker/src/algorithms/certain_zero/search_strategy/linear_constraints.rs

```diff
@@ -0,0 +1,596 @@
+use crate::game_structure::lcgs::ast::{BinaryOpKind, Expr, ExprKind, Identifier, UnaryOpKind};
+use crate::game_structure::lcgs::ir::symbol_table::SymbolIdentifier;
+use std::collections::HashMap;
+use std::fmt::{Display, Formatter};
+
+/// All comparison operators
+#[derive(Hash, Eq, PartialEq, Copy, Clone, Debug)]
+pub enum ComparisonOp {
+    Equal,
+    NotEqual,
+    Less,
+    LessOrEq,
+    Greater,
+    GreaterOrEq,
+}
+
+impl ComparisonOp {
+    fn negated(&self) -> ComparisonOp {
+        match self {
+            ComparisonOp::Equal => ComparisonOp::NotEqual,
+            ComparisonOp::NotEqual => ComparisonOp::Equal,
+            ComparisonOp::Less => ComparisonOp::GreaterOrEq,
+            ComparisonOp::LessOrEq => ComparisonOp::Greater,
+            ComparisonOp::Greater => ComparisonOp::LessOrEq,
+            ComparisonOp::GreaterOrEq => ComparisonOp::Less,
+        }
+    }
+}
+
+impl Display for ComparisonOp {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            ComparisonOp::Equal => write!(f, "="),
+            ComparisonOp::NotEqual => writeln!(f, "!="),
+            ComparisonOp::Less => write!(f, "<"),
+            ComparisonOp::LessOrEq => write!(f, "<="),
+            ComparisonOp::Greater => write!(f, ">"),
+            ComparisonOp::GreaterOrEq => write!(f, ">="),
+        }
+    }
+}
+
+impl From<ComparisonOp> for minilp::ComparisonOp {
+    fn from(op: ComparisonOp) -> Self {
+        match op {
+            ComparisonOp::Equal | ComparisonOp::NotEqual => minilp::ComparisonOp::Eq,
+            ComparisonOp::Less | ComparisonOp::LessOrEq => minilp::ComparisonOp::Le,
```

Comment:
Wont this become a problem for the linear program, that the greater/less comparisons are squished into one? Or maybe this is handled in the linear programming code by subtracting/adding 1? (have not read that part yet)

Reply:
If you look in `linear_programming_search.rs:128-132` you will find that NotEqual's are actually skipped right now.
```
// We skip not-equal constraints since the minilp cannot handle them.
// They don't restrict the solution space in a meaning way anyway.
```
NotEqual's are true everywhere, except on the hyperplane (line), so they are basically true everywhere.
As you mention, we could simulate it with OR, but then we also risk doubling the amount of optimization problems.

Reply:
This comment was meant not about the Equal/NotEqual, but about the other two, Less/LessOrEq into just Le, and Greater/GreaterOrEq into Ge. We lose some precision in this, yes? x < 5 is not the same expression as x <= 5, but in the linear program wont these two expression become the same constraint?

Reply:
And we could add a comment about being able to handle the NotEqual, but we choose not to?

Reply:
> This comment was meant not about the Equal/NotEqual, but about the other two,

I see 🤔 Well, again, by adding an OR we risk doubling the amount of linear problems that need to be solved. And the difference between a `x < 5` and `x <= 5` solution is very small in terms of distance.

Reply:
Thats true, but we could handle the cases for `<=` and `>=` by respectively adding/subtracting one to the constant. Such that `x <= 5` becomes a `x < 6` constraint, and `x >= 5` becomes a `x > 4` constraint?

Reply:
Hold up. Just realised something. `minilp::ComparisonOp::Le` is for less *or equal*, and `minilp::ComparisonOp::Ge` greater *or equal*.

Reply:
Well that's rude of them